### PR TITLE
Adds knock off component

### PR DESCRIPTION
**Pull request by AnturK**
:cl:
add: You can enforce no smoking zones with a disarm aimed at the smokers mouth.
/:cl:

Should i add  this to anything  else ?

### CODE REVIEW

**Comment by nicbn** [comment]
Add this to hats too

File: code/modules/mob/living/carbon/human/species.dm [hunk 1]
```diff
@@ -1213,10 +1213,12 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 		return 1
 	else
 		user.do_attack_animation(target, ATTACK_EFFECT_DISARM)
-
+
```
Comment (praisenarsie) [inline code/modules/mob/living/carbon/human/species.dm]: whitespace?

**Comment by Potato-Masher** [comment]
This might require more balance discussion, but this might be an interesting idea to apply to other limbs to redesign disarm. 

Target Head: Knock their hat off.
Target Arms: Smack an item out of the respective hand you are targeting.
Target Legs/Chest: Push/Trip em over as per usual.

Might be useful for advanced interaction. But, because balence, someone is going to have to approve these ideas.

File: code/datums/components/knockoff.dm [hunk 2]
```diff
@@ -0,0 +1,46 @@
+//Items with these will have a chance to get knocked off when disarming
+/datum/component/knockoff
+	var/knockoff_chance = 100 //Chance to knockoff
+	var/list/target_zones //Aiming for these zones will cause the knockoff, null means all zones allowed
+	var/list/slots_knockoffable //Can be only knocked off from these slots, null means all slots allowed
+	var/datum/component/redirect/disarm_redirect
+
+/datum/component/knockoff/Initialize(knockoff_chance,zone_override,slots_knockoffable)
+	if(!isitem(parent))
+		CRASH("Knockoff component misuse")
```
Comment (Cyberboss) [inline code/datums/components/knockoff.dm]: . = COMPONENT_INCOMPATIBLE before this line

**Comment by SgtxRootbeer** [comment]
If it applies to hats, it shouldn't apply to hats/helmets that have armor values

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/datums/components/knockoff.dm [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+		return
+	if(target_zones && !(zone in target_zones))
+		return
+	if(prob(knockoff_chance))
+		if(wearer.dropItemToGround(I))
+			wearer.visible_message("<span class='warning'>[attacker] knocks off [wearer] [I.name]!</span>","<span class='userdanger'>[attacker] knocks off your [I.name]!</span>")
+
+/datum/component/knockoff/proc/OnEquipped(mob/living/carbon/human/H,slot)
+	if(!istype(H))
+		return
```
[the remark this answers is not in the document]
Comment (AnturK) [reply code/datums/components/knockoff.dm]: Simple animals with hands.

File: code/datums/components/knockoff.dm [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+/datum/component/knockoff/proc/Knockoff(mob/living/attacker,zone)
+	var/obj/item/I = parent
+	var/mob/living/carbon/human/wearer = I.loc
+	if(!istype(wearer))
+		return
+	if(target_zones && !(zone in target_zones))
+		return
+	if(prob(knockoff_chance))
+		if(wearer.dropItemToGround(I))
+			wearer.visible_message("<span class='warning'>[attacker] knocks off [wearer] [I.name]!</span>","<span class='userdanger'>[attacker] knocks off your [I.name]!</span>")
```
[the remark this answers is not in the document]
Comment (AnturK) [reply code/datums/components/knockoff.dm]: Uh early return on what

**Comment by Xhuis** [comment]
Budget insuls should get this, too. Get it? Because they're *knockoff?*

File: code/datums/components/knockoff.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (Cyberboss) [reply code/datums/components/knockoff.dm]: if(!...)
  return

on the two ifs

File: code/datums/components/knockoff.dm (hunk 3 above)
[the remark this answers is not in the document]
Comment (Cyberboss) [reply code/datums/components/knockoff.dm]: :|

File: code/datums/components/knockoff.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (AnturK) [reply code/datums/components/knockoff.dm]: Reads worse for negligible gain to me but alright

File: code/datums/components/knockoff.dm (hunk 3 above)
[the remark this answers is not in the document]
Comment (AnturK) [reply code/datums/components/knockoff.dm]: I mean technically nothing prevents this from working on these, slots just won't exists so i might as well remove this.

File: code/datums/components/knockoff.dm [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	if(!istype(wearer))
+		return
+	if(target_zones && !(zone in target_zones))
+		return
+	if(!prob(knockoff_chance))
+		return
+	if(!wearer.dropItemToGround(I))
+		return
+
+	wearer.visible_message("<span class='warning'>[attacker] knocks off [wearer] [I.name]!</span>","<span class='userdanger'>[attacker] knocks off your [I.name]!</span>")
```
Comment (Cyberboss) [inline code/datums/components/knockoff.dm]: The grammar here seems wonky for the 3rd person message

**Review by Cyberboss — CHANGES_REQUESTED** [review]
QDEL_NULL disarm_redirect in Destroy

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**